kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/EncryptionDekCache.java

```diff
@@ -39,16 +41,66 @@ public class EncryptionDekCache<K, E> {
 
     public static final int NO_MAX_CACHE_SIZE = -1;
     private CipherSpecResolver cipherSpecResolver;
+    private AtomicLong invalidationCount = new AtomicLong(0);
 
     private record CacheKey<K>(K kek, CipherSpec cipherSpec) {}
 
+    private static class RotateOnceDekContext<K, E> implements DekContext<E> {
+        private final CacheKey<K> key;
+        private final Dek<E> dek;
+        private final EncryptionDekCache<K, E> cache;
+        private final AtomicBoolean hasBeenRotated = new AtomicBoolean(false);
+        private final CompletableFuture<Void> rotated = new CompletableFuture<>();
+
+        private RotateOnceDekContext(CacheKey<K> key, Dek<E> dek, EncryptionDekCache<K, E> cache) {
+            this.key = key;
+            this.dek = dek;
+            this.cache = cache;
+        }
+
+        @Override
+        public CompletionStage<Void> rotate() {
+            if (hasBeenRotated.compareAndSet(false, true)) {
+                rotateOnce();
+            }
+            else {
+                LOGGER.debug("encryption key {} has already been rotated", key);
+            }
+            return rotated;
+        }
+
+        private void rotateOnce() {
+            try {
+                LOGGER.debug("rotating encryption key {}", key);
+                dek.destroyForEncrypt();
```

Comment:
The implementation of destroyForEncrypt appears that it might not actually destroy the key.  What happens in that case?